feat(starknet): firehose trigger filter

xJonathanLEI opened the pull request:
Supersedes #5201.

This PR adds 2 trigger filters to Starknet:

- block trigger that specifies a list of ranges
- event trigger that's capable of filtering by contract address, event topic, _and_ block ranges

When used without turning on the `filters` feature on the Firehose provider, these filters only work on the `graph-node` side - meaning that full blocks are still always streamed to `graph-node`. Filters are applied _only to avoid excessive trigger parsing_. This part was what #5201 enabled.

What's new in this new PR is the additional implementation of Firhose triggers (i.e. `to_firehose_filter`), allowing `graph-node` to send a `transforms` field in its Firehose stream subscription requests when the `filters` feature is enabled for the provider. Note that all blocks are still always streamed - the filters only affects the list of transactions that are included.

`firehose-starknet` has already [implemented](https://github.com/starknet-graph/firehose-starknet/commit/ec0ae1e3df841635923ff15f99150c8f03fbb950) support for filters. This PR has been tested against that implementation and it works as expected - blocks without matching events have empty transactions, and those with matching ones only come with relevant transactions.

mangas commented on chain/starknet/src/adapter.rs:
```diff
@@ -1,27 +1,219 @@
     Chain,
 };
 
+type TopicWithRanges = HashMap<Felt, Vec<BlockRange>>;
```
This type is quite opaque in how it's supposed to work, what does Felt represent here? Elsewhere it seems to be an address, here seems to be a "topic". 

Can you please make sure the Felt type comment covers what is represents? 

How many topics do we expect to represent in these? HashMap is probably overkill if it's only handling a handful of values.

xJonathanLEI replied:
Oh sorry for the confusion. `Felt` is the universal primitive type in Starknet. Everything is a `Felt`: from addresses to contract call params to event logs, literally everything. I will make sure to cover this fact in the `Felt` type comment itself.

As for "topic", probably bad naming but what it means is the first element in any event `keys`, similar to Ethereum event signatures (which are the first element of any log "topics"). I should probably name this `signature` instead?

mangas replied:
maybe something like EventPrefix or EventFirstKey something along those lines, signature could make sense too if that's how it's treated, I think adding rust comments to the types will make that clearer regardless of the name :) I'm terrible at naming things eheh

mangas commented on chain/starknet/src/adapter.rs:
```diff
@@ -1,27 +1,219 @@
+
+#[derive(Default, Clone)]
+pub struct StarknetEventFilter {
+    pub contract_addresses: HashMap<Felt, TopicWithRanges>,
```
Also somewhat related would be great to have test coverage here

mangas commented on chain/starknet/src/adapter.rs:
```diff
@@ -1,27 +1,219 @@
 
     fn to_firehose_filter(self) -> Vec<prost_types::Any> {
-        todo!()
+        // An empty event filter list means that the subgraph is not interested in events at all.
```
can you please add some unit tests here

mangas commented on chain/starknet/src/chain.rs:
```diff
@@ -391,20 +392,61 @@ impl TriggersAdapterTrait<Chain> for TriggersAdapter {
-                            block: shared_block.clone(),
-                            transaction: transaction.clone(),
-                        })
+                    .filter_map(|event| {
```
As mentioned in a different comment, would be great if you could move the matching logic into the filter itself leave this implementation focused on extract the triggers from the block rather than dealing with internals from the filter

mangas replied:
Would also be great if we can have some tests, near has some examples you can follow.

mangas reviewed (changes requested):
See other comments

xJonathanLEI commented:
Pushed changes:

- encapsulated trigger filter logic
- renamed a few type aliases to reduce ambiguity
- added trigger filter matching tests

Notably, since the trigger filter type is now completely encapsulated, it's now rather difficult to construct a `TriggerFilter` instance from `chain.rs`. So I just tested everything inside `adapter.rs` instead.

mangas commented:
As I understand it this integration was taken in a different direction, I'll close the PR for now, re-open if needed. Also if not needed anymore would be great to remove the existing stakrnet code.
